ext/hash/tests/hash_serialize_003.phpt

```diff
@@ -241,16 +17,12 @@ function test_serialization($serial, $hash, $algo) {
     }
 }
 
-foreach ($serializations as $slist) {
-    $algo = $slist[0];
+foreach (hash_algos() as $algo) {
```

Comment:
Any thoughts on this change @kohler? How valuable is it to test for the exact outputs, rather than just the round-trip?

Reply:
Having static serialize strings is a good thing to check against. However, there are variants that seem to require a certain platform/build where it's not quite clear how they are produced and how to produce them. As explained in the description, this mechanism doesn't seem really practical when it comes to extending the test. It'll require to have all platforms and to produce all builds that are mentioned in the test.

We don't expect the properties of the hash context to change too often, but it's really a question of practicability vs testability. Having static payloads is nice, as long as it doesn't become too much burden in updating the test.

Thanks.

Reply:
The serialization strings are important because serializations are supposed to either (1) be portable across implementations, or (2) cause an exception when unserialized on an incompatible system. The check against static strings exposed many instances of unportability and I think they should be kept.

I also think the reason for removing them is not strong. I do not believe that HashContext should have a new `args` element added. Instead, the *hash algorithm* should read information from the `args` into the existing `hash->context` object as necessary. No hash algorithm uses `args` after initialization. That's all you need for murmur, and I think it's cleaner and better long term.

Reply:
That's not how it works. It won't be possible to serialize something that is not in the object properties. It's not consistent not being able to serialize/unserialize one of the algorithms.

I'm all for keeping static strings. If you can document how to update this test, it would be best.

Thanks.

Reply:
And also, there will be sure more use of this mechanism in the future. Existing algorithms can use it, new algorithms that require it will use it, too.

Thanks.

Reply:
I suggest 

* `args` should be passed to the algorithm-specific hash context initialization function, as in your current code.
* But there is no need to store a copy of `args` in `struct _php_hashcontext_object`. Instead, the algorithm-specific initialization functions should save information from `args` into the context as appropriate. Murmur already does this. No hash algorithm uses `args` post-initialization.
* Then there is no need to change `struct _php_hashcontext_object` at all, and in particular no need to change the existing static serialization strings.

The correct storage location for algorithm-specific hash state is the existing opaque context, which is designed for exactly that purpose.

Reply:
There is a need for args to survive, because args can be needed at the initialization time as well as potentially at the finalization time. There can be numerous use cases to offer, where the arguments need to survive serialization, clone and all that things. One can read from the previous conversation, that several things could have been done when a mechanism like this existed.

Properties could need to be added to the hash context object by any other change. Thus, it's not that this test can/should be frozen forever, it's a test issue. There should be a way to have the check as robust as by having static strings, but still it should not block new features. Therefore it seems best would be if you could provide instructions on how to update the test strings, or a way to skip the test if it has no valid test strings for a given platform. I can generate strings easily for amd64 or arm64. It depends on what is the approach to do that.

Thanks.

Reply:
I found those strings by running code locally and by using AppVeyor builds etc. The easiest way to modify those strings for a new format would probably be to decode the strings using base64, modify them as appropriate for your preferred format (by adding the new serialization of an empty hash at the proper location), and then re-encode them using base64. Then add the re-encoded versions back into the test.

That said, and I won't harp on this further, but the args just don't need to survive. The args 
are absolutely useful to `hash_init`, which parses them and stores the results in the opaque context. But they are not available to the `hash_update` and `hash_final` functions (e.g., PHP_SHA256Update, PHP_SHA256Final), so anything meaningful to a hash algorithm would need to go into the context anyway. If a real use case is found for preserving the args, you could change the code later to add them. But there isn't a use case now.

Reply:
That sounds ok, the suggested way to fix it should work, thanks. Lets fix this test, as that's the most convenient option. I'll add some doc to the test on how to update it. This might take time till i've time, though. Will think more to maybe refactor.

```
    [   
        "sha3-512",
        "....", /* 64-bit fast LE */
        "....", /* 32-bit fast LE */
        "...." /* slow */
    ]
```
Those `64-bit fast LE`, `32-bit fast LE` and `slow` refer to the SIMD vs. plain C implementations, i'd guess?

For the other part - there is already a use case. For example, CityHash will add the seed(s) in the final step, not at the start. It is correct that only the context matters, however under circumstances it needs to be populated correctly after unserialise, etc. Another use case, not for the hash algorithm itself, but for the features, is for example to offer a possibility to configure the hash output format. Right now it's only raw or hex, there can be more. Finally, algorithms can be added as an out of tree extension, an inconsistency in the core will matter more there especially if it's crosses core versions. The most forward compatible way seems to just handle this properly from the beginning on.

Thanks.  

Thanks

Reply:
> Those 64-bit fast LE, 32-bit fast LE and slow refer to the SIMD vs. plain C implementations, i'd guess?

Yes. “LE” meant little endian, but testing seemed to show that the format is not endian-sensitive.

> It is correct that only the context matters, however under circumstances it needs to be populated correctly after 
unserialise, etc.

Exactly: the context must be populated correctly after unserialize regardless :)

> Another use case, not for the hash algorithm itself, but for the features, is for example to offer a possibility to configure the hash output format. 

Users might expect the hash output format to be passed as an extra argument to hash_final, as is done currently for hex/binary. Passing the output format at that time allows users to choose output format when they want the hash, and allows (through hash_copy) a single accumulated context to be serialized in different ways.

Reply:
P.S. If you actually plan to use the args throughout the hash process, rather than just at initialization to initialize the context, you would need to change all the hash_update and hash_final signatures to take args, rather than just hash_init. However, this would be worse than simply initializing the context with the args

Reply:
K, i guess we can close this. Thanks for the instructions so far, i'll check to still update the test with the steps once this PR is done.

Thank.